File: progect2/progect2/Core/Evidence/PRMath/PRMath.swift
// SPDX-License-Identifier: LicenseRef-Aether3D-Proprietary
// Copyright (c) 2024-2026 Aether3D. All rights reserved.

//
// PRMath.swift
// Aether3D
//
// PR3 - Unified Math Facade
// All mathematical operations in Core/Evidence/ MUST go through PRMath
//
// RULE: Business logic layer can ONLY import PRMath (not PRMathDouble, PRMathFast, etc.)
// FORBIDDEN: Direct use of Foundation.exp, Darwin.exp, pow, tanh, etc.
//
// PERFORMANCE MODES:
// - canonical (default): PRMathDouble, stable sigmoid
// - fast: PRMathFast, LUT-based (shadow/benchmark only)
// - fixed: PRMathFixed, Q32.32 fixed-point (future)
//

import Foundation

/// Unified math facade for Evidence layer
///
/// RULE: All mathematical operations in Core/Evidence/ MUST go through PRMath
/// FORBIDDEN: Direct use of Foundation.exp, Darwin.exp, pow, tanh, etc.
///
/// PERFORMANCE MODES:
/// - canonical (default): PRMathDouble, stable sigmoid
/// - fast: PRMathFast, LUT-based (shadow/benchmark only)
/// - fixed: PRMathFixed, Q32.32 fixed-point (future)
public enum PRMath {

    // ═══════════════════════════════════════════════════════════════════════
    // MARK: - Core Sigmoid Functions
    // ═══════════════════════════════════════════════════════════════════════

    /// Standard sigmoid: σ(x) = 1 / (1 + e^(-x))
    ///
    /// IMPLEMENTATION:
    /// - Canonical: StableLogistic.sigmoid (piecewise formula)
    /// - Fast: LUTSigmoid.sigmoid (256-entry LUT, shadow only)
    ///
    /// GUARANTEE: No NaN, no Inf, output ∈ (0, 1)
    ///
    /// - Parameter x: Input value
    /// - Parameter context: Tier context (determines implementation)
    /// - Returns: Sigmoid value ∈ (0, 1)
    @inlinable
    public static func sigmoid(_ x: Double, context: TierContext = .forTesting) -> Double {
        switch context.tier {
        case .canonical:
            return PRMathDouble.sigmoid(x)
        case .fast:
            #if PRMATH_FAST
            return PRMathFast.sigmoid(x)
            #else
            // Fallback to Double if Fast not available
            return PRMathDouble.sigmoid(x)
            #endif
        case .fixed:
            #if PRMATH_FIXED
            return PRMathFixed.sigmoid(x)
            #else
            // Fallback to Double if Fixed not available
            return PRMathDouble.sigmoid(x)
            #endif
        }
    }

    /// Sigmoid from threshold with transition width
    ///
    /// FORMULA: sigmoid((value - threshold) / slope)
    /// where slope = transitionWidth / 4.4
    ///
    /// - Parameters:
    ///   - value: Input value
    ///   - threshold: 50% point of sigmoid
    ///   - transitionWidth: Width of transition zone (10% to 90%)
    ///   - context: Tier context
    /// - Returns: Sigmoid value ∈ (0, 1)
    @inlinable
    public static func sigmoid01FromThreshold(
        _ value: Double,
        threshold: Double,
        transitionWidth: Double,
        context: TierContext = .forTesting
    ) -> Double {
        // Compute slope from transition width
        let slope = transitionWidth / 4.4
        let safeSlope = max(slope, 1e-10)  // Avoid division by zero
        let normalized = (value - threshold) / safeSlope
        return sigmoid(normalized, context: context)
    }

    /// Inverted sigmoid (for "lower is better" metrics)
    ///
    /// FORMULA: sigmoid((threshold - value) / slope)
    ///
    /// - Parameters:
    ///   - value: Input value (lower is better)
    ///   - threshold: 50% point of sigmoid
    ///   - transitionWidth: Width of transition zone
    ///   - context: Tier context
    /// - Returns: Sigmoid value ∈ (0, 1)
    @inlinable
    public static func sigmoidInverted01FromThreshold(
        _ value: Double,
        threshold: Double,
        transitionWidth: Double,
        context: TierContext = .forTesting
    ) -> Double {
        let slope = transitionWidth / 4.4
        let safeSlope = max(slope, 1e-10)
        let normalized = (threshold - value) / safeSlope
        return sigmoid(normalized, context: context)
    }

    // ═══════════════════════════════════════════════════════════════════════
    // MARK: - Safe Math Functions
    // ═══════════════════════════════════════════════════════════════════════

    /// Safe exponential: clamps input to prevent overflow
    /// INPUT RANGE: clamped to [-80, 80]
    ///
    /// - Parameter x: Input value
    /// - Returns: exp(x) with clamped input
    @inlinable
    public static func expSafe(_ x: Double) -> Double {
        return PRMathDouble.expSafe(x)
    }

    /// Safe atan2: handles NaN/Inf inputs
    ///
    /// NOTE: This is NOT used in canonical path (zero-trig policy)
    /// Only available for shadow verification
    ///
    /// - Parameters:
    ///   - y: Y component
    ///   - x: X component
    /// - Returns: atan2(y, x) with NaN/Inf handling
    @inlinable
    public static func atan2Safe(_ y: Double, _ x: Double) -> Double {
        // Handle degenerate cases
        guard y.isFinite && x.isFinite else {
            // Return 0 for any non-finite input (deterministic fallback)
            return 0.0
        }
        guard !(y == 0 && x == 0) else {
            // atan2(0, 0) is undefined, return 0 (deterministic)
            return 0.0
        }
        return atan2(y, x)
    }

    /// Safe asin: clamps input to [-1, 1]
    ///
    /// NOTE: This is NOT used in canonical path (zero-trig policy)
    /// Only available for shadow verification
    ///
    /// - Parameter x: Input value
    /// - Returns: asin(clamped x) with NaN/Inf handling
    @inlinable
    public static func asinSafe(_ x: Double) -> Double {
        guard x.isFinite else { return 0.0 }
        let clamped = max(-1.0, min(1.0, x))
        return asin(clamped)
    }

    /// Safe sqrt: returns 0 for negative input
    ///
    /// - Parameter x: Input value
    /// - Returns: sqrt(x) or 0 if x < 0 or non-finite
    @inlinable
    public static func sqrtSafe(_ x: Double) -> Double {
        guard x.isFinite && x >= 0 else { return 0.0 }
        return sqrt(x)
    }

    // ═══════════════════════════════════════════════════════════════════════
    // MARK: - Utility Functions
    // ═══════════════════════════════════════════════════════════════════════

    /// Clamp to [0, 1]
    ///
    /// - Parameter x: Input value
    /// - Returns: Clamped value ∈ [0, 1]
    @inlinable
    public static func clamp01(_ x: Double) -> Double {
        return max(0.0, min(1.0, x))
    }

    /// Clamp to arbitrary range
    ///
    /// - Parameters:
    ///   - x: Input value
    ///   - lo: Lower bound
    ///   - hi: Upper bound
    /// - Returns: Clamped value ∈ [lo, hi]
    @inlinable
    public static func clamp(_ x: Double, _ lo: Double, _ hi: Double) -> Double {
        return max(lo, min(hi, x))
    }

    /// Check if value is usable (finite and not NaN)
    ///
    /// - Parameter x: Input value
    /// - Returns: true if value is finite
    @inlinable
    public static func isUsable(_ x: Double) -> Bool {
        return x.isFinite
    }
}

File: progect2/progect2/aether_cpp/src/evidence/pr_math.cpp
// SPDX-License-Identifier: LicenseRef-Aether3D-Proprietary
// Copyright (c) 2024-2026 Aether3D. All rights reserved.

#include "aether/evidence/pr_math.h"

#include <cmath>
#include <limits>

namespace aether {
namespace evidence {

// ═══════════════════════════════════════════════════════════════════════
// StableLogistic
// ═══════════════════════════════════════════════════════════════════════

double StableLogistic::exp_safe(double x) noexcept {
    // Clamp to [-80, 80] to prevent overflow (exp(80) ~ 5.5e34, well within double)
    if (x > 80.0) x = 80.0;
    if (x < -80.0) x = -80.0;
    return std::exp(x);
}

double StableLogistic::sigmoid(double x) noexcept {
    // Handle non-finite inputs deterministically
    if (std::isnan(x)) return 0.5;
    if (std::isinf(x)) return x > 0.0 ? 1.0 : 0.0;

    // Clamp to prevent overflow
    if (x > 80.0) x = 80.0;
    if (x < -80.0) x = -80.0;

    // Piecewise formula for numerical stability:
    // For x >= 0: 1 / (1 + exp(-x))  -- exp(-x) in (0, 1], no overflow
    // For x < 0:  exp(x) / (1 + exp(x)) -- exp(x) in (0, 1), no overflow
    if (x >= 0.0) {
        const double e = std::exp(-x);
        return 1.0 / (1.0 + e);
    } else {
        const double e = std::exp(x);
        return e / (1.0 + e);
    }
}

double StableLogistic::log_sigmoid(double x) noexcept {
    // log(σ(x)) = -log(1 + exp(-x)) = -softplus(-x)
    // For x >= 0: -log(1 + exp(-x)), exp(-x) is small, stable
    // For x < 0: x - log(1 + exp(x)), avoids log(0)
    if (!std::isfinite(x)) {
        if (x != x) return -std::numeric_limits<double>::infinity(); // NaN
        return x > 0.0 ? 0.0 : -std::numeric_limits<double>::infinity();
    }
    if (x >= 0.0) {
        double em = std::exp(-x);
        return -std::log1p(em); // log1p for precision when em is small
    } else {
        double ep = std::exp(x);
        return x - std::log1p(ep); // x + log(σ(x)/exp(x))
    }
}

double StableLogistic::log_complement_sigmoid(double x) noexcept {
    // log(1 - σ(x)) = log(σ(-x)) = log_sigmoid(-x)
    return log_sigmoid(-x);
}

// ═══════════════════════════════════════════════════════════════════════
// PRMath
// ═══════════════════════════════════════════════════════════════════════

double PRMath::sigmoid(double x) noexcept {
    return StableLogistic::sigmoid(x);
}

double PRMath::log_sigmoid(double x) noexcept {
    return StableLogistic::log_sigmoid(x);
}

double PRMath::log_complement_sigmoid(double x) noexcept {
    return StableLogistic::log_complement_sigmoid(x);
}

double PRMath::softplus(double x) noexcept {
    // softplus(x) = log(1 + exp(x))
    // For x > 20: return x (exp(x) dominates, log(exp(x)) = x)
    // For x < -20: return exp(x) (log(1+tiny) ≈ tiny)
    // Otherwise: log1p(exp(x))
    if (x > 20.0) return x;
    if (x < -20.0) return std::exp(x);
    return std::log1p(std::exp(x));
}

double PRMath::sigmoid01_from_threshold(double value, double threshold,
                                         double transition_width) noexcept {
    if (!is_usable(value) || !is_usable(threshold)) return 0.5;
    if (!is_usable(transition_width) || transition_width <= 0.0) {
        // Hard step
        return value >= threshold ? 1.0 : 0.0;
    }
    // Map to sigmoid: when value == threshold -> input is 0 -> sigmoid returns 0.5
    const double scaled = (value - threshold) / transition_width;
    return StableLogistic::sigmoid(scaled);
}

double PRMath::sigmoid_inverted01_from_threshold(double value, double threshold,
                                                   double transition_width) noexcept {
    return 1.0 - sigmoid01_from_threshold(value, threshold, transition_width);
}

double PRMath::exp_safe(double x) noexcept {
    return StableLogistic::exp_safe(x);
}

double PRMath::atan2_safe(double y, double x) noexcept {
    if (std::isnan(y) || std::isnan(x)) return 0.0;
    if (std::isinf(y) || std::isinf(x)) return 0.0;
    return std::atan2(y, x);
}

double PRMath::asin_safe(double x) noexcept {
    if (std::isnan(x)) return 0.0;
    // Clamp to valid domain
    if (x > 1.0) x = 1.0;
    if (x < -1.0) x = -1.0;
    return std::asin(x);
}

double PRMath::sqrt_safe(double x) noexcept {
    if (std::isnan(x) || x < 0.0) return 0.0;
    return std::sqrt(x);
}

double PRMath::clamp01(double x) noexcept {
    if (std::isnan(x)) return 0.0;
    if (x < 0.0) return 0.0;
    if (x > 1.0) return 1.0;
    return x;
}

double PRMath::clamp(double x, double lo, double hi) noexcept {
    if (std::isnan(x)) return lo;
    if (x < lo) return lo;
    if (x > hi) return hi;
    return x;
}

bool PRMath::is_usable(double x) noexcept {
    return std::isfinite(x);
}

// ═══════════════════════════════════════════════════════════════════════
// QuantizerQ01
// ═══════════════════════════════════════════════════════════════════════

int64_t QuantizerQ01::quantize(double value) noexcept {
    // Clamp to valid range (defensive)
    double clamped = value;
    if (clamped < 0.0) clamped = 0.0;
    if (clamped > 1.0) clamped = 1.0;
    // Round half away from zero (deterministic, matches Swift .toNearestOrAwayFromZero)
    return static_cast<int64_t>(std::round(clamped * kScale));
}

double QuantizerQ01::dequantize(int64_t q) noexcept {
    return static_cast<double>(q) / kScale;
}

bool QuantizerQ01::are_close(int64_t a, int64_t b, int64_t tolerance) noexcept {
    int64_t diff = a - b;
    if (diff < 0) diff = -diff;
    return diff <= tolerance;
}

}  // namespace evidence
}  // namespace aether
